### PR TITLE
[bluetooth.daikinmadoka] Daikinmadoka binding initial contribution

### PR DESCRIPTION
**Pull request by cpmeister**
Signed-off-by: Benjamin Lafois <benjamin@lafois.com>

This PR was made on behalf of @blafois.

### CODE REVIEW

File: bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/BRC1HUartProcessor.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    /**
+     * In the unlikely event of messages arrive in wrong order, this comparator will sort the queue
+     */
+    private Comparator<byte[]> chunkSorter = (byte[] m1, byte[] m2) -> m1[0] - m2[0];
+
+    ///////////////////////////////////////////////////////////////////////////////////////////////
+
+    private final Logger logger = LoggerFactory.getLogger(BRC1HUartProcessor.class);
+
+    private PriorityBlockingQueue<byte[]> uartMessages = new PriorityBlockingQueue<>(10, chunkSorter);
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/BRC1HUartProcessor.java:44]: It should be noted that this class iterator elements in order, it only guarantees that polling from it will return the smallest element. I think what you might want to use instead is a ConcurrentSkipListSet which does has iteration ordering guarantees.

File: bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/BRC1HUartProcessor.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+        int expectedChunks = (firstMessageInQueue[1] / 19) + (firstMessageInQueue[1] % 19 > 0 ? 1 : 0);
+        if (expectedChunks != this.uartMessages.size()) {
+            return false;
+        }
+        {
+            // Check that we have every single ID
+            byte[][] messages = this.uartMessages.toArray(new byte[][] {});
+            int expected = 0;
+            for (byte[] m : messages) {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/BRC1HUartProcessor.java:85]:
```suggestion
            int expected = 0;
            for (byte[] m : this.uartMessages) {
```

File: bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/model/MadokaMessage.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+        // Message Length - Computed in the end
+        request = ArrayUtils.add(request, (byte) 0);
+        request = ArrayUtils.add(request, (byte) 0);
+
+        // Command ID, coded on 3 bytes
+        request = ArrayUtils.add(request, (byte) 0);
+        ByteBuffer bb = ByteBuffer.allocate(2);
+        bb.putShort((short) command.getCommandId());
+        request = ArrayUtils.addAll(request, bb.array());
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/model/MadokaMessage.java:64]: Please don't use a ByteBuffer for such a trivial operation. Just use bitwise operators instead.
Another option which would look a bit better would be to build the message bytes through a DataOutputStream.
```
ByteArrayOutputStream output = new ByteArrayOutputStream();
DataOutputStream request = new DataOutputStream(output);
request.writeByte(0);
request.writeByte(0);
request.writeByte(0);
request.writeShort(command.getCommandId());
...
byte[] ret = output.toByteArray();
ret[1] = (byte) (request.length -1);
return ret;
```

File: bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/model/MadokaMessage.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        }
+        if (msg.length < 1) {
+            return null;
+        }
+        if (msg[0] != msg.length) {
+            return null;
+        }
+
+        MadokaMessage m = new MadokaMessage();
+        m.messageId = ByteBuffer.wrap(msg, 2, 2).getShort();
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/model/MadokaMessage.java:99]: Please use bitwise operations instead.

File: bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/model/MadokaProperties.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.bluetooth.daikinmadoka.internal.model;
+
+/**
+ *
+ * @author blafois
+ *
+ */
+public class MadokaProperties {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/model/MadokaProperties.java:20]: Please add `@NonNullByDefault`

File: bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/model/MadokaMessage.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+import org.openhab.binding.bluetooth.daikinmadoka.internal.model.commands.BRC1HCommand;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ *
+ * @author blafois
+ *
+ */
+public class MadokaMessage {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/model/MadokaMessage.java:30]: Please add `@NonNullByDefault`

File: bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/model/MadokaSettings.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.openhab.binding.bluetooth.daikinmadoka.internal.model.MadokaProperties.FAN_SPEED;
+import org.openhab.binding.bluetooth.daikinmadoka.internal.model.MadokaProperties.OPERATION_MODE;
+
+/**
+ *
+ * @author blafois
+ *
+ */
+public class MadokaSettings {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/model/MadokaSettings.java:25]: Please add `@NonNullByDefault`

File: bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/DaikinMadokaHandlerFactory.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+ */
+@NonNullByDefault
+@Component(configurationPid = "binding.bluetooth.daikinmadoka", service = ThingHandlerFactory.class)
+public class DaikinMadokaHandlerFactory extends BaseThingHandlerFactory {
+
+    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Collections.singleton(THING_TYPE_BRC1H);
+
+    private final Logger logger = LoggerFactory.getLogger(DaikinMadokaHandlerFactory.class);
+
+    ///////////////////////////////////////////////////////////////////////////////////////////////
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/DaikinMadokaHandlerFactory.java:46]: please remove all these comment separator blocks from all files

File: bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/model/MadokaValue.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+package org.openhab.binding.bluetooth.daikinmadoka.internal.model;
+
+import java.nio.ByteBuffer;
+
+/**
+ *
+ * @author blafois
+ *
+ */
+public class MadokaValue {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/model/MadokaValue.java:22]: Please add `@NonNullByDefault`

File: bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/model/MadokaValue.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+    private int id;
+    private int size;
+    private byte[] rawValue;
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/model/MadokaValue.java:26]: If this value can be null then you will have to add the following annotation.
```suggestion
    private byte @Nullable [] rawValue;
```

File: bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/model/commands/BRC1HCommand.java [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.eclipse.jdt.annotation.Nullable;
+
+/**
+ *
+ * @author blafois
+ *
+ */
+public abstract class BRC1HCommand {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/model/commands/BRC1HCommand.java:28]: Please add `@NonNullByDefault`

File: bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/model/commands/GetFanspeedCommand.java [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+import org.openhab.binding.bluetooth.daikinmadoka.internal.model.MadokaProperties.FAN_SPEED;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ *
+ * @author blafois
+ *
+ */
+public class GetFanspeedCommand extends BRC1HCommand {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/model/commands/GetFanspeedCommand.java:28]: Please add `@NonNullByDefault`

File: bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/model/commands/GetFanspeedCommand.java [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+            MadokaMessage mm = MadokaMessage.parse(response);
+
+            this.coolingFanSpeed = FAN_SPEED.valueOf(mm.getValues().get(0x20).getRawValue()[0]);
+            this.heatingFanSpeed = FAN_SPEED.valueOf(mm.getValues().get(0x21).getRawValue()[0]);
+
+            logger.debug("coolingFanSpeed: {}", coolingFanSpeed);
+            logger.debug("heatingFanSpeed: {}", heatingFanSpeed);
+
+            listener.receivedResponse(this);
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/model/commands/GetFanspeedCommand.java:61]: The reason that I passed the executor in this `handleResponse` method was so that any eventual listener callback would be executed asynchronously on it. I had a very good reason to do it as well since the bluetooth notification callback would start dropping messages if you didn't handle the notification as quick as possible. If enough messages got dropped eventually the whole notification framework stopped sending notifications causing my binding to stop getting updates from sensors.

So in short, just make sure that any channel updates are done on a separate thread from your notifications.
```suggestion
            executor.execute(() -> listener.receivedResponse(this));
```

File: bundles/org.openhab.binding.bluetooth.daikinmadoka/src/test/java/org/openhab/binding/bluetooth/daikinmadoka/handler/DummyThing.java [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+
+    @Override
+    public @Nullable String setProperty(String name, @Nullable String value) {
+        // TODO Auto-generated method stub
+        return null;
+    }
+
+    @Override
+    public void setProperties(Map<@NonNull String, @NonNull String> properties) {
+        // TODO Auto-generated method stub
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.daikinmadoka/src/test/java/org/openhab/binding/bluetooth/daikinmadoka/handler/DummyThing.java:152]: Please remove all these auto-generated comments

File: bundles/org.openhab.binding.bluetooth.daikinmadoka/src/test/java/org/openhab/binding/bluetooth/daikinmadoka/handler/UARTQueueTester.java [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+package org.openhab.binding.bluetooth.daikinmadoka.handler;
+
+/**
+ *
+ * @author blafois
+ *
+ */
+public class UARTQueueTester {
+
+    // @Test
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.daikinmadoka/src/test/java/org/openhab/binding/bluetooth/daikinmadoka/handler/UARTQueueTester.java:22]: What is wrong with all these tests? I would use them if possible but if you can't use them then just get rid of them.

File: bundles/org.openhab.binding.bluetooth.daikinmadoka/src/test/java/org/openhab/binding/bluetooth/daikinmadoka/internal/MadokaMessageTest.java [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+
+        byte[] testMsg = new byte[] { 0x00, 0x10 };
+
+        assertEquals(1, MadokaMessage.expectedMessageChunks(testMsg));
+
+    }
+
+    @Test
+    public void testExpectedMessageChunks1() {
+
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.daikinmadoka/src/test/java/org/openhab/binding/bluetooth/daikinmadoka/internal/MadokaMessageTest.java:42]: please remove all the excess blank lines in this file

File: bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/DaikinMadokaBindingConstants.java [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+    public static final String CHANNEL_ID_OPERATION_MODE = "operationMode";
+    public static final String CHANNEL_ID_FAN_SPEED = "fanSpeed";
+    public static final String CHANNEL_ID_SETPOINT = "setpoint";
+    public static final String CHANNEL_ID_HOMEKIT_CURRENT_HEATING_COOLING_MODE = "homekitCurrentHeatingCoolingMode";
+    public static final String CHANNEL_ID_HOMEKIT_TARGET_HEATING_COOLING_MODE = "homekitTargetHeatingCoolingMode";
+
+    /**
+     * BLUETOOTH UUID (service + chars)
+     */
+    public static final String SERVICE_UART_UUID = "2141E110-213A-11E6-B67B-9E71128CAE77";
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/DaikinMadokaBindingConstants.java:48]: Why not just make them UUID instances to begin with so you don't have to parse them later on?
```suggestion
    public static final UUID SERVICE_UART_UUID = UUID.fromString("2141E110-213A-11E6-B67B-9E71128CAE77");
```

File: bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/BRC1HUartProcessor.java (hunk 1 above)
Comment (blafois) [reply bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/BRC1HUartProcessor.java:44]: Thanks for spotting this - I actually did the unit test with pull - and refactored the code then iterating over. Will fix.

File: bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/DaikinMadokaHandlerFactory.java (hunk 8 above)
Comment (blafois) [reply bundles/org.openhab.binding.bluetooth.daikinmadoka/src/main/java/org/openhab/binding/bluetooth/daikinmadoka/internal/DaikinMadokaHandlerFactory.java:46]: oh noooooo !!

**Comment by blafois** [comment]
Hello Connor ! I have updated everything according to your comments!

**Comment by blafois** [comment]
Fixed everything!

**Comment by cpmeister** [comment]
@blafois is expected to resubmit this PR.
